posts/hoxonstudios/cirular-dependencies-microservices.md

```diff
@@ -0,0 +1,73 @@
+## Leaking domain knowledge
+
+A very imperceptible way of creating circular dependencies between services is sending requests with Ids, value objects, or any other kind of knowledge that belongs to the client service domain.
+
+### Example
+
+Suppose we have these services:
+
+- ProductService
+- InvoiceService
+
+Inside our **ProductService** we will surely have a **Product** with each entity having a **ProductId**. In this example **ProductService** will be sending a request to **InvoiceService** to generate an invoice that could be something like this:
+
+```text
+POST /invoice
+{ productId: 3, quantity: 1, ... }
+```
+
+In this situation **ProductService** has to know that some invoice service exists to send the request. However, our **InvoiceService** will also be aware of the existence of some **Product** entity that is in the domain of the caller. A good way to verify if we are creating a circular dependency is thinking about the **InvoiceService** as a separated entity. If we completely remove the **ProductService**, will this request still make sense as it is? The obvious answer will be 'No'.
+
+To solve this, first we have to decide which service will depend on the other, say Product on Invoice. After we decide that **InvoiceService** will not depend on **ProductService**, we need to think about the former in isolation. Do we really need to receive a **ProductId** to validate the existence of the product, get its unit price and description, and then generate the invoice? Or do we just need a **ProductCode** to display in the invoice? Most of the time the answer will be that receiving the **ProductCode**, **UnitPrice** and **Description** in the request is good enough. Trusting the information received from services behind our control is crucial to avoid unnecessary validations.
+
+```text
+POST /invoice
+{
+  productCode: 'A123',
+  description: 'Nice product',
+  unitPrice: 99.9,
+  quantity: 1,
+  ...
+}
+```
+
+This is just a simple example but keep in mind that there are a lot of ways to leak domain knowledge to other services without us noticing.
+
+## Choreography
+
+This communication pattern seems to be very common in Event-Driven systems. In it, services emit events into a common Event Bus so others can react to them to complete a complex distributed process.
+
+### Example
+
+- **UserService** emits a *UserRegisteredEvent*
+- **CreditService** receives that *UserRegisteredEvent* and validates the credit of that person
+- **CreditService** emits a *CreditValidatedEvent*
+- **UserService** receives that *CreditValidatedEvent* and finishes the user creation process
+
+Some people may think that, as we are using events to communicate  between services, we are completelly decoupling them. That is only partially true, because the only kind of coupling we are removing is the temporal coupling. On the other hand, we are now creating a circular dependency between them because events, as requests, are part of the API of our services.
```

Comment:
```suggestion
Some people may think that, as we are using events to communicate between services, we are completelly decoupling them. That is only partially true, because the only kind of coupling we are removing is the temporal coupling. On the other hand, we are now creating a circular dependency between them because events, as requests, are part of the API of our services.
```